[WIP] Add density based CNN clustering to sklearn.cluster

janjoswig opened the pull request:
#### What does this implement?

This PR adds common-nearest-neighbours (CNN) clustering to the collection of available algorithms. CNN is a density-based cluster algorithm closely related to DBSCAN, which is already included in scikit-learn. The two algorithms provide a different notion of the local point density, yielding in general similar results. While in DBSCAN the local point density around a point is estimated by the number of neighbouring points with respect to a radius cutoff, in CNN the number of common neighbouring points between two points (i.e. the neighbourhood intersection) is used. Depending on the data situation at hand one can be more favourable to use than the other. In my research, I am getting nice results when clustering Molecular Dynamics trajectories with the CNN algorithm. It would be interesting to see, how the algorithm performs compared to DBSCAN and other density-based algorithms in different applications.

The provided CNN implementation closely follows the present realization of DBSCAN, adding little to no additional complexity to the project. I think it would be great to have the two algorithms available side by side, where one can choose flexibly between them and compare the outcome.

The CNN algorithm has been described in the following research papers:
 
   - B. Keller, X. Daura, W. F. van Gunsteren "Comparing Geometric and
    Kinetic Cluster Algorithms for Molecular Simulation Data" J. Chem.
    Phys., 2010, 132, 074110.

  - O. Lemke, B.G. Keller "Density-based Cluster Algorithms for the
    Identification of Core Sets" J. Chem. Phys., 2016, 145, 164104.

  - O. Lemke, B.G. Keller "Common nearest neighbor clustering - a
    benchmark" Algorithms, 2018, 11, 19.

I am also working right now on a stand-alone package implementing the CNN algorithm alongside handy features used in the context of Molecular Dynamics. The package is available at [github.com/janjoswig/CNN](https://github.com/janjoswig/CNN).

#### Any other comments?

I have opened this PR in the [WIP] state to see if this has any chance of being accepted. If there is a general interest in this addition, I am happy to work on tests and documentation including benchmark and comparison.

janjoswig committed:
WIP: Added _cnn_inner extension to setup.py

janjoswig committed:
ENH: Added _cnn_inner.pyx and _cnn.py

janjoswig committed:
TEST: doctest for _cnn.py

janjoswig committed:
FIX: _cnn.py docstring

janjoswig committed:
ENH: Added examples/cluster/plot_cnn.py

glemaitre commented:
I think this work does not fulfill the inclusion criteria (200+ citations). However, a scikit-learn-contrib package (if you want to manage the package yourself) or inclusion to `scikit-learn-extra` (if you just want to only care about the method without packaging) should be possible.

Maybe @rth that follows more `scikit-learn-extra` could know if this would be a good idea.

glemaitre commented:
We would still need some testing in all cases which is missing right now.

rth commented:
Yes, please feel free to make a PR with this implementation and tests to https://github.com/scikit-learn-contrib/scikit-learn-extra

janjoswig committed:
FIX: Account for self-counting in neighbourhoods.

  - Points are their own neighbours in the neighbourhoods
 (`neighbors.NearestNeighbors`)
  - `min_samples` + 1 in general necessary to account for
  the extra point
  - `min_samples` + 2 necessary when checking the similarity criterion
  (neighbouring points without real common neighbours automatically
  have themselves as two fake common neighbours)

janjoswig committed:
TEST: Added test for CNN clustering

 - Adapts tests from `_test_dbscan`

janjoswig commented:
Ok, great. I adapted the tests for the DBSCAN implementation. So I am closing this PR and move over to `scikit-learn-extra`.  I really appreciate it.
